elixir/apps/api/test/api/client/channel_test.exs

```diff
@@ -333,18 +350,18 @@ defmodule API.Client.ChannelTest do
                password: _,
                type: _,
                username: _
-             } = relay_view1
+             } = relay_view
 
-      Domain.Relays.Presence.untrack(self(), "presences:relays:#{relay.id}", relay.id)
+      Domain.Relays.Presence.untrack(self(), "presences:relays:#{relay1.id}", relay1.id)
 
       assert_push "relays_presence", %{
-        disconnected_ids: [relay_id],
+        disconnected_ids: [relay1_id],
         connected: [relay_view1, relay_view2]
       }
 
-      assert relay_view1.id == relay.id
-      assert relay_view2.id == relay.id
-      assert relay_id == relay.id
+      assert relay_view1.id == relay2.id
+      assert relay_view2.id == relay2.id
+      assert relay1_id == relay1.id
     end
```

Comment:
The property that we are interested in, and which currently doesn't hold is:

```suggestion
        assert relay_view1.id != relay1_id
        assert relay_view2.id != relay1_id
    end
```

It might be worth adding 1 more test that represents staging / prod more closely where we have ~10 relays, disconnect them all one by one, connect new ones, capture all relays presence messages and assert that none of the disconnected ones ever get returned as connected.

That test doesn't need to capture which specific ones of the newly connected ones do get returned so you don't need to simulate the selection logic. We use property-based testing connlib extensively and it helped us find a lot of these subtle bugs :)

Reply:
> Isn't
> 
> ```
> assert relay_view1.id == relay2.id
> ```
> 
> is the same as
> 
> ```
> assert relay_view1.id != relay1_id
> ```
> 
> because we know that `relay2.id != relay1_id`?

In this test setup, yes. If you'd use more inputs, modelling which ones is actually returned gets more difficult and obscures the test. In connlib, we don't care which one we get but the property of not receiving a disconnected one is really important. What I am suggesting is IMO simply a more intentional assertion of what we want. It is your codebase, so up to you :)